# PS3 Bonus Question

## 1. How many Spam Emails are there?

In [1]:
def count_spam(path = "D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data", folders = ['enron1','enron2','enron3']):
    import os
    from pathlib import Path
    #path = Path("D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data")
    path = Path(path)
    spam_count = 0
    for folder in folders:
        file_names = os.listdir(path / folder / 'spam')
        for file_name in file_names:
            if 'spam.txt' in file_name:
                spam_count += 1
    print('There are', spam_count, 'spam emails in', folders)

In [2]:
count_spam()

There are 4492 spam emails in ['enron1', 'enron2', 'enron3']


## 2. Structure the email data from the 2 directories into 1 dataframe with columns: Status, Subject, Body.

#### Specify the folders that will be included and the path leading to the folders.

In [3]:
folders = ['enron1','enron2','enron3']
status = ['spam','ham']
path1 = "D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\{folder}\\{status}"

#### Create the paths containing all the desired emails.

In [4]:
from itertools import product
directories = list(product(folders, status))
directories

[('enron1', 'spam'),
 ('enron1', 'ham'),
 ('enron2', 'spam'),
 ('enron2', 'ham'),
 ('enron3', 'spam'),
 ('enron3', 'ham')]

In [5]:
paths = []
for i in directories:
    paths.append(path1.format(folder=i[0],status=i[1]))
paths

['D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron2\\spam',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron2\\ham',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron3\\spam',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron3\\ham']

#### List all the files in each `path`.

In [6]:
import glob
file_paths = [glob.glob(pathname=path + '\\*') for path in paths]

#### Now we have all the file paths of the emails in the specified folders.

In [7]:
file_paths[0][0:5]

['D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0006.2003-12-18.GP.spam.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0008.2003-12-18.GP.spam.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0017.2003-12-18.GP.spam.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0018.2003-12-18.GP.spam.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0026.2003-12-18.GP.spam.txt']

In [8]:
file_paths[1][0:5]

['D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0001.1999-12-10.farmer.ham.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0002.1999-12-13.farmer.ham.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0003.1999-12-14.farmer.ham.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0004.1999-12-14.farmer.ham.txt',
 'D:\\Cadetship\\IntroDS\\HW2\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0005.1999-12-14.farmer.ham.txt']

#### Having acquired the file paths, we can now extract the status, subject and message body of the emails.

In [9]:
subjects = []
bodies = []
status = []

# for each path
for file_path in file_paths:
    # for each file in path
    for i in file_path:
        # for status
        if 'spam.txt' in i:
            status.append('spam')
        elif 'ham.txt' in i:
            status.append('ham')
        # extract the subject from email
        with open(i,'r',errors='ignore') as f:
            subject = f.readlines()[0]
            subject = subject[9:]
            subjects.append(subject)
        # extract body message from email
        with open(i,'r',errors='ignore') as f:
            body = f.readlines()
            body = body[1:]
            bodies.append(body)
            
# join each sub-elements of the bodies list
bodies = [' '.join(i) for i in bodies]

print('Status has', len(status), 'elements.')
print('Subject has', len(subjects), 'elements.')
print('Body has', len(bodies), 'elements.')

Status has 16537 elements.
Subject has 16537 elements.
Body has 16537 elements.


#### Now, use the lists to create a dataframe.

In [10]:
import pandas as pd

In [11]:
enron = pd.DataFrame(list(zip(status,subjects,bodies)), columns=['Status','Subject','Body'])
enron.sample(10)

Status                                       Subject  \
15501    ham         enron center south technology watch\n   
10469    ham                                 sevil yamin\n   
12820    ham            re : corporate culture committee\n   
2715     ham                           re : saudi arabia\n   
1741     ham                                  meter 7266\n   
3874     ham                             re : meter 1517\n   
9776     ham                                   iris mack\n   
13883    ham                 rrevised weekly deal report\n   
15413    ham          fw : dynegy master netting / sithe\n   
12198   spam  please kindly respond ( dr . frank aluko )\n   

                                                    Body  
15501  welcome to enron center south . as you have pr...  
10469  anne ,\n vasant sent this information to norma...  
12820  i love committees and this sounds like a real ...  
2715   daren :\n i ' ll follow - up on this , but i '...  
1741   daren - can you set up a deal for meter 7266 ?...  
3874   daren - after checking with mips , you are cor...  
9776   vince : i received a phone call yesterday afte...  
13883  please see the corrections to done deals ( int...  
15413  carol st . clair\n eb 4539\n 713 - 853 - 3989 ...  
12198  confidential\n 2 adefobi street ,\n ikoyi isla...

## 3. Build a Naive Bayes classifier to classify whether emails are spam or not.

#### Import libraries

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [13]:
enron.sample(5)

Status                                        Subject  \
7248    ham  re : argentina power & gas market modelling\n   
5058    ham         fw : sitara eol bridge problem today\n   
7460    ham                                   power 2000\n   
2526    ham                                  adios . . .\n   
1827    ham                     cdnow order confirmation\n   

                                                   Body  
7248  team ,\n for me it  % s ok . let us know the ...  
5058  fyi > > > we were also monitoring the eol to s...  
7460  power 2000 - may 8 - 11\n ?\n please note to h...  
2526  as most of you have probably heard already , i...  
1827  dear daren ,\n thank you for shopping at cdnow...

In [14]:
enron.shape

(16537, 3)

#### Create `Spam` column from the `Status` column and delete `Status` column.

In [15]:
enron['Spam'] = [1 if i == 'spam' else 0 for i in enron['Status']]
enron.drop('Status', inplace=True, axis=1)

In [16]:
enron.sample(5)

Subject  \
1006   never get a traffic ticket again ! - ia 55 m\n   
10070     el paso statement on california situation\n   
3493                                re : meter 9707\n   
5611            urgent paypal security notification\n   
13707                                        louise\n   

                                                    Body  Spam  
1006   have you been caught by a red light camera yet...     1  
10070  in case you haven ' t already seen this :\n el...     0  
3493   daren ,\n fyi .\n bob\n - - - - - - - - - - - ...     0  
5611   security center advisory !\n we recently notic...     1  
13707  hope all is well . you can reach me by phone o...     0

#### Split data for training and testing.

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
train_X, test_X, train_y, test_y = train_test_split(enron["Body"], enron["Spam"], test_size=0.3, random_state=0)

#### Import libraries for text analysis

In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import GaussianNB

In [20]:
vectorizer = TfidfVectorizer(stop_words='english')
tfidf_train_X = vectorizer.fit_transform(train_X)

#### Train the model

In [21]:
classifier = GaussianNB()
classifier.fit(tfidf_train_X.toarray(), train_y)

GaussianNB()

## 4. What is the longest ham email?

#### Create a dataframe containing all ham emails.

In [22]:
hams = enron[enron['Spam'] == 0][['Body']]

#### Create a column containing the lengths of each emails.

In [23]:
hams['Lengths'] = [len(i) for i in hams['Body']]
hams.drop('Body',inplace=True,axis=1)

#### Get the index of the longest ham email.

In [24]:
longest_ham = hams.idxmax()
print(longest_ham)

Lengths    15750
dtype: int64


#### Print the longest ham email.

In [25]:
enron.iloc[longest_ham,1]

15750    enron : a wake - up call\n the wall street jou...
Name: Body, dtype: object

#### The longest ham email has `230120` characters.

In [26]:
hams.max()

Lengths    230120
dtype: int64

## 5. What is the accuracy of your model?

In [27]:
from sklearn.model_selection import cross_val_score

In [28]:
tfidf_test_X = vectorizer.transform(test_X)
print(tfidf_test_X.shape)
scores = cross_val_score(classifier, tfidf_test_X.toarray(), test_y, cv=5)
acc = scores.mean()
print("Accuracy: %0.2f percent" % (acc *100))

(4962, 80072)
Accuracy: 92.52 percent


**Model accuracy is `92.52%`**

## 6. Include the Subject in the analysis of the emails, does the accuracy/performance of the model increase?

#### Create a new column containing the concatenation of email `Subject` and `Body`.

In [29]:
enron['Total'] = [sub + bod for (sub, bod) in zip(enron.iloc[:,0],enron.iloc[:,1])]

In [30]:
enron.head(5)

Subject  \
0  dobmeos with hgh my energy level has gone up !...   
1        your prescription is ready . . oxwq s f e\n   
2                            get that new car 8434\n   
3                              await your response\n   
4  coca cola , mbna america , nascar partner with...   

                                                Body  Spam  \
0  introducing\n doctor - formulated\n hgh\n huma...     1   
1  low cost prescription medications\n soma , ult...     1   
2  people nowthe weather or climate in any partic...     1   
3  dear partner ,\n we are a team of government o...     1   
4  stock\n profile\n about\n company\n investment...     1   

                                               Total  
0  dobmeos with hgh my energy level has gone up !...  
1  your prescription is ready . . oxwq s f e\nlow...  
2  get that new car 8434\npeople nowthe weather o...  
3  await your response\ndear partner ,\n we are a...  
4  coca cola , mbna america , nascar partner with...

In [31]:
train_X, test_X, train_y, test_y = train_test_split(enron["Total"], enron["Spam"], test_size=0.3, random_state=0)

In [32]:
vectorizer = TfidfVectorizer(stop_words='english')
tfidf_train_X = vectorizer.fit_transform(train_X)

In [33]:
classifier = GaussianNB()
classifier.fit(tfidf_train_X.toarray(), train_y)

GaussianNB()

In [34]:
tfidf_test_X = vectorizer.transform(test_X)
print(tfidf_test_X.shape)
scores = cross_val_score(classifier, tfidf_test_X.toarray(), test_y, cv=5)
acc = scores.mean()
print("Accuracy: %0.2f percent" % (acc *100))

(4962, 81140)
Accuracy: 93.31 percent


**The accuracy improved by `0.79%`**